kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md

```diff
@@ -18,301 +18,587 @@ module EVM-OPTIMIZATIONS-LEMMAS [kore]
 endmodule
 
 module EVM-OPTIMIZATIONS [kore]
-    imports EVM
-    imports EVM-OPTIMIZATIONS-LEMMAS
-    imports INT-SIMPLIFICATION
+  imports EVM
+  imports EVM-OPTIMIZATIONS-LEMMAS
+  imports INT-SIMPLIFICATION
 
-rule <kevm>
-       <k>
-         ( #next[ PUSHZERO ] => . ) ...
-       </k>
-       <schedule>
-         SCHED
-       </schedule>
-       <ethereum>
-         <evm>
-           <callState>
-             <wordStack>
-               ( WS => 0 : WS )
-             </wordStack>
-             <pc>
-               ( PCOUNT => ( PCOUNT +Int 1 ) )
-             </pc>
-             <gas>
-               ( GAVAIL => ( GAVAIL -Gas Gbase < SCHED > ) )
-             </gas>
-             ...
-           </callState>
-           ...
-         </evm>
-         ...
-       </ethereum>
-       ...
-     </kevm>
-  requires ( Gbase < SCHED > <=Gas GAVAIL )
-   andBool ( #sizeWordStack( 0 : WS ) <=Int 1024 )
-    [priority(40)]
+  rule
+    <kevm>
+      <k>
+        ( #next[ PUSHZERO ] => . ) ...
+      </k>
+      <useGas>
+        false
+      </useGas>
+      <ethereum>
+        <evm>
+          <callState>
+            <wordStack>
+              ( WS => 0 : WS )
+            </wordStack>
+            <pc>
+              ( PCOUNT => ( PCOUNT +Int 1 ) )
+            </pc>
+            ...
+          </callState>
+          ...
+        </evm>
+        ...
+      </ethereum>
+      ...
+    </kevm>
+  requires ( #sizeWordStack( 0 : WS ) <=Int 1024 )
+  [priority(40)]
```

Comment:
I guess these rules can be de-duplicated by doing something like this instead:

```suggestion
  rule
    <kevm>
      <k>
        ( #next[ PUSHZERO ] => . ) ...
      </k>
      <useGas>
        USE_GAS
      </useGas>
      <ethereum>
        <evm>
          <callState>
            <wordStack>
              ( WS => 0 : WS )
            </wordStack>
            <pc>
              ( PCOUNT => ( PCOUNT +Int 1 ) )
            </pc>
            <gas>
              GAVAIL => #if USE_GAS #then GAVAIL -Gas Gbase < SCHED > #else GAVAIL #fi
            </gas>
            ...
          </callState>
          ...
        </evm>
        ...
      </ethereum>
      ...
    </kevm>
  requires ( #sizeWordStack( 0 : WS ) <=Int 1024 )
  [priority(40)]
```

Then we wouldn't be increasing the number of rules in the higher-priority bucket (40), so that we don't introduce more rules the backend has to try.

Reply:
Yes, I think this would work well if the value of `<useGas>` was never symbolic.

Reply:
Ah, actually, we also have a `requires Gbase < SCHED > <=Gas GAVAIL` when `USE_GAS` is true, and nothing if it's false.